nixie/nixie.go

```diff
@@ -0,0 +1,239 @@
+	KvmOptions dcp.Options `group:"Hetec DCP Serial client"`
+}
+
+func (options Options) Make() (*Nixie, error) {
+	var nixie = Nixie{
+		options: options,
+	}
+
+	if err := nixie.init(options); err != nil {
+		return nil, err
+	}
+
+	return &nixie, nil
+}
+
+type Nixie struct {
+	options Options
+
+	//Livepin is HIGH if the kvm channel is live
+	livePin embd.DigitalPin
+
+	spiBus     embd.SPIBus
+	kvmConsole int
+	kvmTallies [4]bool
+
+	kvmChan   chan int
+	tallyChan chan tally.State
+	closeChan chan bool
+	waitGroup sync.WaitGroup
+}
+
+func (nixie *Nixie) init(options Options) error {
+	fmt.Printf("Init GPIO\n")
+
+	if err := embd.InitGPIO(); err != nil {
+		return fmt.Errorf("embd.InitGPIO: %v", err)
+	}
+
+	if err := embd.InitSPI(); err != nil {
+		panic(err)
+	}
+
+	fmt.Printf("Intialize SPI\n")
+	nixie.spiBus = embd.NewSPIBus(embd.SPIMode0, 0, 50, 8, 100)
+
+	nixie.Send(nixie_no_number | nixie_white)
+	nixie.Send(nixie_no_number | nixie_white)
+	nixie.Send(nixie_no_number | nixie_white)
+
+	if err := embd.InitGPIO(); err != nil {
+		return fmt.Errorf("embd.InitGPIO: %v", err)
+	}
+
+	if options.LivePin == "" {
+
+	} else if pin, err := embd.NewDigitalPin(options.LivePin); err != nil {
+		return fmt.Errorf("embd.NewDigitalPin %v: %v", options.LivePin, err)
+
+		// Writing as "out" defaults to initializing the value as low.
+	} else if err := pin.SetDirection(embd.Out); err != nil {
+		return fmt.Errorf("pin.SetDirection %v: %v", options.LivePin, err)
+	} else {
+		nixie.livePin = pin
+	}
+
+	nixie.kvmConsole = 5
+	nixie.kvmChan = make(chan int)
+
+	nixie.closeChan = make(chan bool)
+
+	return nil
+}
+
+func (nixie *Nixie) RegisterTally(t *tally.Tally) {
+	nixie.tallyChan = make(chan tally.State)
+	nixie.waitGroup.Add(1)
+
+	go nixie.run()
+
+	t.Register(nixie.tallyChan)
+}
+
+func (nixie *Nixie) close() {
+	defer nixie.waitGroup.Done()
+
+	log.Printf("Nixie: Close pins and SPI bus..")
+
+	if nixie.livePin != nil {
+		nixie.livePin.Close()
+	}
+
+	// Turn off the nixie tube and release the spi bus
+	nixie.Send(nixie_no_number | nixie_led_off)
+	nixie.spiBus.Close()
+	embd.CloseSPI()
+}
+
+func (nixie *Nixie) updateTally(state tally.State) {
+	log.Printf("Nixie: Update tally State:")
+	fmt.Printf("\tKVM tallies: ")
+	for id := tally.ID(1); id < 5; id++ {
+		var pinState = false
+
+		if status, exists := state.Tally[id]; !exists {
+			// missing tally state for pin
+		} else {
+			if status.Status.Program {
+				pinState = true
+			}
+		}
+		fmt.Printf("%d: %t ", id, pinState)
+		nixie.kvmTallies[id-1] = bool(pinState)
+	}
+	fmt.Printf("\n")
+}
+
+func (nixie *Nixie) run() {
+	defer nixie.close()
+
+	go nixie.listenKvm()
+
+	// Initialize the nixie tube
+	log.Printf("Initialize nixie tube")
+	nixie.Send(nixie_numbers[0] | nixie_yellow)
+
+	log.Printf("Entering message loop")
+	for {
+		select {
+		case nixie.kvmConsole = <-nixie.kvmChan:
+			log.Printf("KVM console: %d", nixie.kvmConsole)
+		case state := <-nixie.tallyChan:
```

Comment:
Ideally this should check for `state, open := <-nixie.tallyChan` to shutdown if the tally is stopped.

Not a real problem because atm Tally will never close the `tallyChan`.